### PR TITLE
[rllib] Add deterministic test to gpu

### LINKED ISSUE #18245: [rllib] Reproduce issue: with the same seed, multiple runs still have different performance
### What is the problem?

I found multiple runs for PPO still have different performance even we set the same seed.
How can we obtain the exact same result with the same seed?

Current strategy of rllib sets the platform seed as follows:

https://github.com/ray-project/ray/blob/db9b5f142d8c91a872bff00735eb8574dd374b6e/rllib/evaluation/rollout_worker.py#L77-L112

I attached the results of two runs with both `seed=0`

Snipaste_2021-08-31_15-01-50
Snipaste_2021-08-31_15-02-17

*Ray version and other system information (Python version, TensorFlow version, OS):*
- Ray-dev.2.0  (commit 8136d2912b195a7a6b84f675e18165f873b2138e)
- Pytorch: 1.8.0+cu111
- CUDA driver: NVIDIA-SMI 470.63.01    Driver Version: 470.63.01    CUDA Version: 11.4  
- python：3.8
- ubuntu 20.04
### Reproduction (REQUIRED)
1. install the latest rllib (commit 8136d2912b195a7a6b84f675e18165f873b2138e)
2. run the `cartpole_lstm` example on two gpus
```python
import argparse
import os

from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole
from ray.rllib.utils.test_utils import check_learning_achieved

parser = argparse.ArgumentParser()
parser.add_argument(
    "--run",
    type=str,
    default="PPO",
    help="The RLlib-registered algorithm to use.")
parser.add_argument("--num-cpus", type=int, default=0)
parser.add_argument(
    "--framework",
    choices=["tf", "tf2", "tfe", "torch"],
    default="torch",
    help="The DL framework specifier.")
parser.add_argument("--eager-tracing", action="store_false")
parser.add_argument("--use-prev-action", action="store_true")
parser.add_argument("--use-prev-reward", action="store_true")
parser.add_argument(
    "--as-test",
    action="store_true",
    help="Whether this script should be run as a test: --stop-reward must "
    "be achieved within --stop-timesteps AND --stop-iters.")
parser.add_argument(
    "--stop-iters",
    type=int,
    default=200,
    help="Number of iterations to train.")
parser.add_argument(
    "--stop-timesteps",
    type=int,
    default=100000,
    help="Number of timesteps to train.")
parser.add_argument(
    "--stop-reward",
    type=float,
    default=150.0,
    help="Reward at which we stop training.")

if __name__ == "__main__":
    import ray
    from ray import tune

    args = parser.parse_args()
    cwd_path = os.getcwd()
    print('cwd path', cwd_path)
    logdir = cwd_path + '/log'
    print(logdir)
    if not os.path.isdir(logdir):
        os.mkdir(logdir)
    ray.init(logging_level='DEBUG')

    configs = {
        "PPO": {
            "num_sgd_iter": 5,
            "model": {
                "vf_share_layers": True,
            },
            "vf_loss_coeff": 0.0001,
        },
        "IMPALA": {
            "num_workers": 2,
            "num_gpus": 0,
            "vf_loss_coeff": 0.01,
        },
    }

    config = dict(
        configs[args.run],
        **{
            "env": StatelessCartPole,
            # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
            "num_gpus": 1,
            "num_workers": 1,
            "num_gpus_per_worker": (2 - 1) / (1+1),
 
            "grad_clip": 0.5,#tune.grid_search([1, 5]),

            "seed": 0,
            "model": {
                "use_lstm": True,
                "lstm_cell_size": 256,
                "lstm_use_prev_action": args.use_prev_action,
                "lstm_use_prev_reward": args.use_prev_reward,
            },
            "framework": args.framework,
            # # Run with tracing enabled for tfe/tf2?
            # "eager_tracing": args.eager_tracing,
        })

    stop = {
        "training_iteration": args.stop_iters,
        "timesteps_total": args.stop_timesteps,
        "episode_reward_mean": args.stop_reward,
    }


    results = tune.run(args.run, config=config, verbose=3, checkpoint_freq=50, local_dir=logdir, stop=stop)

    if args.as_test:
        check_learning_achieved(results, args.stop_reward)
    ray.shutdown()


```

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### LINKED ISSUE #18784: [Bug] `linux://rllib:examples/rnnsac_stateless_cartpole` is slow
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

`linux://rllib:examples/rnnsac_stateless_cartpole` is slow. Please consider splitting it, speeding it up, or utilizing GPU.

### Reproduction script

```
linux://rllib:examples/rnnsac_stateless_cartpole  ml          653.33    
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by avnishn**
## Why are these changes needed?

Move Deterministic training ex to Multi-GPU Tests
    
Moves all deterministic_training.py examples to multi gpu runners on the CI, adding test coverage for issues like #18245

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by sven1977 — CHANGES_REQUESTED** [review]
Hey @avnishn . Thanks for the PR and for looking into enhancing our GPU CI tests!
Just a few comments and smaller changes.

**Review by simon-mo — APPROVED** [review]
do we still need the `-multi_gpus` if your new tests only matches thoses tag?

**Review by sven1977 — APPROVED** [review]
Looks great, thanks for the remaining fixes!

**Review by richardliaw — APPROVED** [review]
nice! looks good on my side (for setup.py). someone else should review carefully.
